src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.TryGet.cs

```diff
@@ -0,0 +1,172 @@
+
+using System.Buffers.Text;
+using System.Diagnostics;
+
+namespace System.Text.Json
+{
+    public ref partial struct Utf8JsonReader
+    {
+        // Reject any invalid UTF-8 data rather than silently replacing.
+        private static readonly UTF8Encoding s_utf8Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
+
+        /// <summary>
+        /// Reads the next JSON token value from the source as a <see cref="string"/>.
+        /// Returns true if the UTF-8 encoded token value can be successfully transcoded
+        /// to a UTF-16 string.
+        /// </summary>
+        /// <remarks>
+        /// This method never returns false. If the JSON token is a string and the data is valid UTF-8, the cast is always successful.
+        /// Otherwise, it throws.
+        /// </remarks>
+        /// <exception cref="InvalidCastException">
+        /// Thrown if trying to get the value of JSON token that is not a string
+        /// (i.e. JsonTokenType.String, JsonTokenType.PropertyName, or JsonTokenType.Comment).
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// Thrown if invalid byte sequence are detected while transcoding.
+        /// </exception>
+        public bool TryGetValueAsString(out string value)
+        {
+            if (TokenType != JsonTokenType.String && TokenType != JsonTokenType.PropertyName && TokenType != JsonTokenType.Comment)
+            {
+                throw ThrowHelper.GetInvalidCastException_ExpectedString(TokenType);
+            }
+
+            // TODO: https://github.com/dotnet/corefx/issues/33292
+            value = s_utf8Encoding.GetString(ValueSpan);
+            return true;
+        }
+
+        /// <summary>
+        /// Reads the next JSON token value from the source as a <see cref="bool"/>.
+        /// Returns true if the entire UTF-8 encoded token value can be successfully 
+        /// parsed to a <see cref="bool"/> value.
+        /// </summary>
+        /// <remarks>
+        /// This method never returns false. If the JSON token is a boolean, the cast is always successful.
+        /// Otherwise, it throws.
+        /// </remarks>
+        /// <exception cref="InvalidCastException">
+        /// Thrown if trying to get the value of JSON token that is not a boolean (i.e. JsonTokenType.True or JsonTokenType.False).
+        /// </exception>
+        public bool TryGetValueAsBoolean(out bool value)
+        {
+            if (TokenType == JsonTokenType.True)
+            {
+                Debug.Assert(ValueSpan.Length == 4);
+                value = true;
+            }
+            else if (TokenType == JsonTokenType.False)
+            {
+                Debug.Assert(ValueSpan.Length == 5);
+                value = false;
+            }
+            else
+            {
+                throw ThrowHelper.GetInvalidCastException_ExpectedBoolean(TokenType);
+            }
+            return true;
+        }
```

Comment:
I'm confused by the `Try` prefix now.  In what situations do we return `false` from these `Try` methods?

Reply:
I followed @bartonjs's suggestion to make this change: https://github.com/dotnet/corefx/pull/33216#discussion_r231328722

> What does this method do when the precision of the number doesn't fit in double?

> For AsnReader my TryGet things are

> * Success: Return true
> * Right data type, value doesn't fit: Return false
> * Wrong data type, you should have checked: Throw.

> I like that one because the token type is something the caller can already reason about, and then the return false is about the content.

> In this method you're mixing "wrong token type" and "Utf8Parser.TryParse said no", which means I don't know what to do as a caller (the rule for Try methods is all "return false" paths should have the same caller response, and anything other than the one optimized false should throw -- sometimes it feels like "never throw", but there's a lot of nuance)

If we decide to revert this, I would go back to returning false in all cases and following the TryGet pattern.

Reply:
> which means I don't know what to do as a caller

The caller can disambiguate if they need to by checking the token type first.  I don't see this as a case of the `Try` bool being used to mean two different things, e.g. I ask it to parse "9876543210" as an Int32, and I ask it to parse "987654321d" as an Int32... I don't see why I should expect false for the first case and an exception for the latter.  int.TryParse, for example, doesn't throw if it hits something non-numeric, it returns false.

Reply:
Framework Design Guidelines (2ed), 7.5.2 (Try-Parse Pattern)

> When using this pattern, it is important to define the try functionality in **strict terms**.  If the member fails for any reason other than the well-defined try, the member **must still throw** a corresponding exception.

(Emphasis mine)

@KrzysztofCwalina said that what this means is something like  "every return false should imply one recovery action".  In the case of Span-filling methods it means "there's not enough space".  My suggestion here is that false means "this is a JSON number, but not one that fits in Int32" (so try Int64 or double), but that isn't the same recovery that you'd take if the current token type was PropertyName.

Reply:
> My suggestion here is that false means "this is a JSON number, but not one that fits in Int32"

And mine is: "this cannot be parsed as an Int32" :)  I'm not sure why it's valuable to distinguish whether it's too big to be an Int32 or contains non-digits by returning false for one and throwing an exception for the other.

Reply:
That just adds conceptual overhead that doesn't buy me anything.

The reader has a TryGetValueAsInt32 method.  Conceptually why should that behave any differently from e.g. `int.TryParse(ValueSpan)`?  I'm asking it to try to parse `Value` as an Int32.

Reply:
The excerpt from design guidelines should be interpreted as follows: normally APIs should throw exceptions when the API errors. Such APIs should throw a different exception type for each error that can be handled independently and the same exception for all errors that are handled in the same way. TryXxx pattern allows to replace one and only one of these independently handled error conditions with false return value, i.e. do not replace what otherwise would be more than one exception type with false return.

Reply:
Right.  I think our disagreement here stems from what does it actually mean for there to be "different" errors.  From my perspective, we've already decided what that categorization means, with APIs like int.TryParse, BinaryPrimitives.TryRead, etc., returning true if the data could be parsed as the target type and otherwise false.  I'm not sure why we'd make a different categorization here.

Reply:
To me the difference is the reader state.  My model is the state of "I think I'm looking at a property" means calling GetValueAsInt32() would be `InvalidOperationException`, and "this is too big" would be `FormatException` (let's pretend for a moment that int.Parse unified OverflowException and FormatException).

My feeling is that the caller has a program error if they're calling TryGetValueAsInt32 when the reader is on PropertyName, hence exception.

I'm okay with either way, and really it comes down to how we think someone would use it... which we might not actually know at this juncture.

Reply:
> I think our disagreement here stems from what does it actually mean for there to be "different" errors.

Whoever thinks (@bartonjs?) there should be two separate errors should write a code sample showing how these two errors could be handled differently/independently in some real scenario. If such sample cannot be produced, we should treat these two conditions as one error.

Reply:
```C#
private static double GetValueAsPercentage(this Utf8JsonReader reader, int max)
{
    if (reader.TryGetValueAsInt32(out int intVal))
        return 100.0 * intVal / max;
    if (reader.TryGetValueAsInt64(out long longVal))
        return 100.0 * longVal / max;
    if (reader.TryGetValueAsDouble(out double doubleVal))
        return 100.0 * doubleVal / max;

    // Vaporware extension method
    if (reader.TryGetValueAsBigInteger(out BigInteger bigVal))
    {
        // Some BigInteger math goes here.
    }

    // Remaining cases:
    // * The number is too precise (trim it, we're doing division)
    // * The number is too close to zero (return +/- Epsilon)
    // * The number is unfathomably large (1e999999999999999999999999)
    throw new OverflowException("Value doesn't fit");
}
```

This code either needs to start with a "am I a number" (and throw), end with "am I an number?" (and throw something different), or just count on the framework to throw; or it's throwing the wrong exception.

If the expectation is that this code should have checked the TokenType first, then really that means that we don't expect anyone to call TryGetValueAsInt32 without checking that it's a number; so why _wouldn't_ it throw in that state?  (At which point they don't need to, so I go back in time, so I don't, et cetera)

But I'm having trouble envisioning code that does the other thing... gets false because it's a string and has a sensible answer for moving on to TryGetValueAsString.

Reply:
Talked with @bartonjs offline and here is the conclusion:
- Rename `TryGetValueAsX` to `TryGetXValue` since 'As' implies a quick re-interpretation, while we are actually transforming (and it isn't quick - requires encoding/parsing/etc.). This makes the .NET type conversion more explicit so it is clearer to the caller when we throw on mismatched token type.
- Change `TryGetStringValue` and `TryGetBooleanValue` to just `GetStringValue` and `GetBooleanValue` which throw and return void since these will never return false.
- Keep the "number"-based APIs as TryGetXValue returning bool, but these throw InvalidOperationException for mismatched token type.
- To support round-tripping (in the sense that someone can create a JSON payload from reading all the tokens of an existing valid JSON using these APIs and that JSON payload should remain valid. We want to make sure that when the new JSON payload is passed in again to the Utf8JsonReader, it shouldn't throw a JsonReaderException for invalid JSON), if Utf8Parser.TryParse(double) returns true with a `double.PositiveInfinity` or `double.NegativeInfinity` for really large numbers (based on IEEE compliance work that @tannergooding  is doing), we do not return that value, but instead return false.


In the interest of time, and to unblock this PR, I am going to go with this approach for now.

We could have a follow up discussion and change the semantics based on new API additions (GetXValue version of the TryGetXValue APIs and TryGetBigIntegerValue) during an API review: https://github.com/dotnet/corefx/issues/33320.

Reply:
> but these throw InvalidOperationException for mismatched token type

I don't agree with the decision.  I've heard the arguments, and I still do not believe this should diverge from an expectation that these methods are any different `Try`-wise than `int.TryParse` and `BinaryPrimitives.TryRead`. But I'm not going to stand in the way of the PR.